Build failure: mldonkey

## Steps To Reproduce

Steps to reproduce the behavior:

1. build mldonkey

## Build log

<details>

<summary>Build Log (tail)</summary>

```
src/utils/lib/charsetstubs.c: In function 'raise_error':
src/utils/lib/charsetstubs.c:171:15: warning: assignment discards 'const' qualifier from pointer target type [8;;https://gcc.gnu.org/onlinedocs/gcc-14.2.0/gcc/Warning-Options.html#index-Wdiscarded-qualifiers-Wdiscarded-qualifiers8;;]
  171 |     closure_f = caml_named_value("charset_error");
      |               ^
src/utils/lib/charsetstubs.c:173:13: warning: "raise_constant" is deprecated: use "caml_raise_constant" instead
  173 |   raise_constant(*closure_f);
      |             ^~~~~~~~~~~~~~~~~                                                      
src/utils/lib/charsetstubs.c: In function 'ml_iconv':
src/utils/lib/charsetstubs.c:1207:23: error: passing argument 2 of 'iconv' from incompatible pointer type [8;;https://gcc.gnu.org/onlinedocs/gcc-14.2.0/gcc/Warning-Options.html#index-Wincompatible-pointer-types-Wincompatible-pointer-types8;;]
 1207 |     return iconv (cd, inbuf, inbytes_left, outbuf, outbytes_left);
      |                       ^~~~~
      |                       |
      |                       const char **
In file included from src/utils/lib/charsetstubs.c:45:
/nix/store/kj8hbqx4ds9qm9mq7hyikxyfwwg13kzj-glibc-2.40-36-dev/include/iconv.h:49:54: note: expected 'char ** restrict' but argument is of type 'const char **'
   49 | extern size_t iconv (iconv_t __cd, char **__restrict __inbuf,
      |                                    ~~~~~~~~~~~~~~~~~~^~~~~~~
src/utils/lib/charsetstubs.c: In function 'ml_copy_string_len_and_free':
src/utils/lib/charsetstubs.c:1339:13: warning: "alloc_string" is deprecated: use "caml_alloc_string" instead
 1339 |   v = alloc_string (len);
      |             ^~~~~~~~~~~~~                                                      
src/utils/lib/charsetstubs.c: In function 'ml_convert_string':
src/utils/lib/charsetstubs.c:1351:13: warning: "string_length" is deprecated: use "caml_string_length" instead
 1351 |   c_res = ml_convert(String_val(str),string_length(str),
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                         
src/utils/lib/charsetstubs.c: In function 'ml_utf8_validate':
src/utils/lib/charsetstubs.c:1361:13: warning: "string_length" is deprecated: use "caml_string_length" instead
 1361 |   return Val_bool(utf8_validate(SizedString_val(s),NULL));
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                       
src/utils/lib/charsetstubs.c: In function 'ml_locale_charset':
src/utils/lib/charsetstubs.c:1380:13: warning: "copy_string" is deprecated: use "caml_copy_string" instead
 1380 |   return (copy_string ((char*) str));
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~                                        
make: *** [Makefile:5476: src/utils/lib/charsetstubs.o] Error 2
```
</details>

Full log: https://hydra.nixos.org/build/282590211/nixlog/1

Also, `wget` is not detected due to build failure.  These were warnings previously, and now are errors:

<details>

```
wget.c: In function 'main':
wget.c:114:9: error: implicit declaration of function 'close'; did you mean 'pclose'? [8;;https://gcc.gnu.org/onlinedocs/gcc-14.2.0/gcc/Warning-Options.html#index-Wimplicit-function-declaration-Wimplicit-function-declaration8;;]
  114 |         close(f);
      |         ^~~~~
      |         pclose
wget.c:130:6: error: implicit declaration of function 'write'; did you mean 'fwrite'? [8;;https://gcc.gnu.org/onlinedocs/gcc-14.2.0/gcc/Warning-Options.html#index-Wimplicit-function-declaration-Wimplicit-function-declaration8;;]
  130 |      write(fd, buf, strlen(buf));
      |      ^~~~~
      |      fwrite
wget.c:132:20: error: implicit declaration of function 'read'; did you mean 'fread'? [8;;https://gcc.gnu.org/onlinedocs/gcc-14.2.0/gcc/Warning-Options.html#index-Wimplicit-function-declaration-Wimplicit-function-declaration8;;]
  132 |     while((nread = read(fd, buf, 1)) > 0){
      |                    ^~~~
      |                    fread
Cannot compile wget.c
checking for wget... no
```
</details>

## Additional context

The first error looks similar to that addressed in #369191.

There are new versions which may address issues and/or obviate the need for patches:
https://github.com/ygrek/mldonkey/releases

## Metadata

Logs are from Hydra.

## Notify maintainers

@LeSuisse @sternenseemann @wegank

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/applications/networking/p2p/mldonkey/default.nix
+++ b/pkgs/applications/networking/p2p/mldonkey/default.nix
@@ -1,53 +1,64 @@
 {
   lib,
   stdenv,
-  fetchurl,
-  fetchpatch,
+  fetchFromGitHub,
+  autoreconfHook,
+  autoconf-archive,
   ocamlPackages,
+  pkg-config,
   zlib,
 }:
 
 stdenv.mkDerivation rec {
   pname = "mldonkey";
-  version = "3.1.7-2";
+  version = "3.2.1";
 
-  src = fetchurl {
-    url = "https://ygrek.org/p/release/mldonkey/mldonkey-${version}.tar.bz2";
-    sha256 = "b926e7aa3de4b4525af73c88f1724d576b4add56ef070f025941dd51cb24a794";
+  src = fetchFromGitHub {
+    owner = "ygrek";
+    repo = "mldonkey";
+    tag = "release-${lib.replaceStrings [ "." ] [ "-" ] version}";
+    hash = "sha256-Dbb7163CdqHY7/FJY2yWBFRudT+hTFT6fO4sFgt6C/A=";
   };
 
-  patches = [
-    # Fixes C++17 compat
-    (fetchpatch {
-      url = "https://github.com/ygrek/mldonkey/pull/66/commits/20ff84c185396f3d759cf4ef46b9f0bd33a51060.patch";
-      hash = "sha256-MCqx0jVfOaLkZhhv0b1cTdO6BK2/f6TxTWmx+NZjXME=";
-    })
-    # Fixes OCaml 4.12 compat
-    (fetchpatch {
-      url = "https://github.com/ygrek/mldonkey/commit/a153f0f7a4826d86d51d4bacedc0330b70fcbc34.patch";
-      hash = "sha256-/Muk3mPFjQJ48FqaozGa7o8YSPhDLXRz9K1EyfxlzC8=";
-    })
-    # Fixes OCaml 4.14 compat
-    (fetchpatch {
-      url = "https://github.com/FabioLolix/AUR-artifacts/raw/6721c2d4ef0be9a99499ecf2787e378e50b915e9/mldonkey-fix-build.patch";
-      hash = "sha256-HPW/CKfhywy+Km5/64Iok4tO9LJjAk53jVlsYzIRPfs=";
-    })
-  ];
-
-  preConfigure = ''
-    substituteInPlace Makefile --replace '+camlp4' \
-      '${ocamlPackages.camlp4}/lib/ocaml/${ocamlPackages.ocaml.version}/site-lib/camlp4'
+  postPatch = ''
+    substituteInPlace config/Makefile.in \
+      --replace-fail '+camlp4' '${ocamlPackages.camlp4}/lib/ocaml/${ocamlPackages.ocaml.version}/site-lib/camlp4'
   '';
 
   strictDeps = true;
-  nativeBuildInputs = with ocamlPackages; [
-    ocaml
-    camlp4
+
+  nativeBuildInputs = [
+    autoreconfHook
+    autoconf-archive
+    ocamlPackages.camlp4
+    ocamlPackages.findlib
+    ocamlPackages.ocaml
+    pkg-config
   ];
-  buildInputs = (with ocamlPackages; [ num ]) ++ [ zlib ];
+
+  buildInputs = [
+    ocamlPackages.num
+    zlib
+  ];
+
+  preAutoreconf = ''
+    cd config
+  '';
+
+  postAutoreconf = ''
+    cd ..
+  '';
+
+  env =
+    {
+      NIX_CFLAGS_COMPILE = "-Wno-error=implicit-function-declaration";
+    }
+    # https://github.com/ygrek/mldonkey/issues/117
+    // lib.optionalAttrs stdenv.cc.isClang {
+      CXXFLAGS = "-std=c++98";
+    };
 
   meta = {
-    broken = stdenv.hostPlatform.isDarwin;
     description = "Client for many p2p networks, with multiple frontends";
     homepage = "https://github.com/ygrek/mldonkey";
     license = lib.licenses.gpl2Only;

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -14447,7 +14447,7 @@ with pkgs;
   mixxx = qt6Packages.callPackage ../applications/audio/mixxx { };
 
   mldonkey = callPackage ../applications/networking/p2p/mldonkey {
-    ocamlPackages = ocaml-ng.ocamlPackages_4_14_unsafe_string;
+    ocamlPackages = ocaml-ng.ocamlPackages_4_14;
   };
 
   mmex = callPackage ../applications/office/mmex {

